### ISSUE TITLE
How do documents get stored in Xcode?

### ISSUE THREAD

**Issue by sferoze** [martijnwalraven/meteor-ios#12]
Hello,

I successfully integrated the package and was able to login and call methods on the server with `Meteor.loginWithEmail` and `Meteor.callMethodWithName`

But now I am wondering about the integration with Core Data, how to get and display documents

I know you can subscribe to a publication with the code:

```
Meteor.addSubscriptionWithName("notes")
```

Typically when you subscribe to a publication in the browser, the server sends you all the documents which is stored in Minimongo. 

Since we have no Minimongo here, where are the documents stored once you subscribe, and how do you retrieve them? 

How to search and retrieve documents we subscribe too?

Thanks!

**Comment by sferoze** [comment]
Oh yah and also how can I observe a particular collection and have a function called whenever the collection is updated?

I am familiar with this javascript DDP framework: https://github.com/oortcloud/node-ddp-client

In this framework we can subscribe to a collection and then setup and observer on the collection. Anytime it changes we are sent the new documents. 

I am wondering how to do this with this framework.

**Comment by martijnwalraven** [comment]
There is no real Minimongo, but documents are stored in memory. If you use Core Data, you can use its querying and change notification mechanisms as usual. This is currently the preferred approach and it offers many benefits (e.g. relationship support).

Without Core Data, there is no way to really query collections at the document level (you either get a document by ID or fetch all documents), but there is some more information on this ‘low-level’ API in the Wiki:
https://github.com/martijnwalraven/meteor-ios/wiki/Usage-without-Core-Data

**Comment by sferoze** [comment]
Hi Martin, thanks! I am close to getting it working. I was able to subscribe and println the array of documents I got back.

But I want to do core data integration like you mentioned but am confused on how to set it up. I have read over your basic usage instructions and I am here:

```
You'll need to set up a managed object model in Xcode, as you would normally do when using Core Data. Entities correspond to collections (with an automatic singular-plural name mapping a la Rails), and properties to fields. The default mapping will often work fine, but you can specify a different collectionName or fieldName as userInfo in the model editor. All types of relationships – one-to-one, one-to-many and many-to-many – are supported. By default, references are stored in documents on both sides (two-way referencing). But you can specify storage = false as userInfo for a relationship end in the model editor.
```

So is all I need to do just configure the .xcdatamodeld file with all the collections and fields that match my mongoDB collections and fields? 

Then is all you need to do something like this:

```
let fetchRequest = NSFetchRequest(entityName: "Notes")
let fetchedResults = managedContext.executeFetchRequest(fetchRequest, error: &error) as [NSManagedObject]?
```
So basically I when I subscribe I should expect the Entities I setup to automatically be filled up? 
Then just query like normal with Core Data stuff?

**Comment by martijnwalraven** [comment]
Yes, that's the idea. Entity names are usually singular though (so 'Note' in your example). By default, this maps to a plural collection name (e.g. 'notes').

**Comment by sferoze** [comment]
Hi Martin,

I am not able to use your package to have it automatically save to Core Data. I wonder what I am doing wrong. Here is what works.

I can successfully login, and subscribe to a publications. I can println the array of documents I have received back from the server. This is all good, but when I try and println the fetched entity from Core Data it is empty. 

The collection I am getting back is called 'sets'. I made a Core Data entity called 'Set' 

In this entity I only created 2 attributed, 'name', and 'createdAt'. Even though there are more attributes in each document, I only added 2. Do we need to include all attributes for it to work? 

Here is the code I am using in my View Controller:

```
import UIKit
import Foundation
import Meteor

class ViewController: UIViewController, SubscriptionLoaderDelegate {
    
    let appDelegate = UIApplication.sharedApplication().delegate as AppDelegate
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        
        appDelegate.Meteor.loginWithEmail("testios@testios.com", password: "oakhill") { (error) -> Void in
            if error != nil {
                println(error.localizedFailureReason)
            } else {
                println("logging in worked")
                self.callSetsMethod()
            }
        }
        
        
    }
    
    func callSetsMethod() {
        self.appDelegate.Meteor.callMethodWithName("getSetsMobile", parameters: nil) { result in
            println(result.0)
        }
        let subscriptionLoader = SubscriptionLoader()
        var subscription = subscriptionLoader.addSubscriptionWithName("mySets")
        self.subscriptionLoader(subscriptionLoader, subscription: subscription, didFailWithError: NSError())
    }
    
    func subscriptionLoader(subscriptionLoader: SubscriptionLoader, subscription: METSubscription, didFailWithError error: NSError) {
        
        subscriptionLoader.delegate = self
        subscriptionLoader.whenReady {
            var sets = self.appDelegate.Meteor.database.collectionWithName("sets")
            var allSets: NSArray = sets.allDocuments
            println(allSets)
            self.getSetsFromLocalStorage()
        }
    }
    
    func fetch(entityName entity: String, sortedBy sorter: NSSortDescriptor?, withPredicate predicate: NSPredicate?) -> AnyObject? {

        let managedContext = appDelegate.managedObjectContext!
        let fetchRequest = NSFetchRequest(entityName: entity)
        
        if let sort: NSSortDescriptor = sorter {
            fetchRequest.sortDescriptors = [sort]
        }
        
        if let pred: NSPredicate = predicate {
            fetchRequest.predicate = pred
        }
        
        var error: NSError?
        
        let fetchedResults =
        managedContext.executeFetchRequest(fetchRequest,
            error: &error) as [NSManagedObject]?
        
        if let results = fetchedResults {
            return results
        } else {
            println("Could not fetch \(error), \(error!.userInfo)")
            return nil
        }
    }
    
    func getSetsFromLocalStorage() {
        
        var tempEntity: AnyObject? = self.fetch(entityName: "Set", sortedBy: nil, withPredicate: nil)
                
        if let entity: AnyObject = tempEntity {
            
            let entityArray = entity as? [NSManagedObject]
            
            if let entityArray = entityArray {
                println(entityArray) // PRINTS AN EMPTY ARRAY, BUT AM EXPECTING SAVED DOCUMENTS FROM SERVER
            }
        }
    }


    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}
```

**Comment by martijnwalraven** [comment]
You don't have to define all attributes, so that's definitely not the problem. How did you define the entity in the Core Data model? Have you set the class name and module name correctly?

**Comment by sferoze** [comment]
hmm well I took a screenshot to show you how I defined the 'Set' Entity. I defined in in the core data file + created a custom swift class for it, as well as made sure to assign the Entity to the class. 

![Image of Xcode](http://feroze.org/screenshots/xcode_set_entity.png)

As you can see I have created the Entity and also assigned the 'Set' class to it. 

This is the code I have in the Set.swift class file:

```
import CoreData

class Set: NSManagedObject {
    @NSManaged var name: String!
}
```

So I think I have set it up correctly. Am I fetching the entity and printing it out correctly?

**Comment by martijnwalraven** [comment]
Ah, I think I know what might be going wrong. From the code you posted, it seems you're setting up your own Core Data stack in the AppDelegate. That'll create a duplicate stack (probably using sqlite storage) that is not connected to the Meteor stack. You should remove all boilerplate Core Data code from the AppDelegate.

Then instead of getting the managed object context from the AppDelegate:
```
let managedContext = appDelegate.managedObjectContext!
```
You simply use:
```
Meteor.mainQueueManagedObjectContext
```

Also, there is no need to use `self.appDelegate.Meteor`. You can just use `Meteor` (e.g. `Meteor.callMethodWithName(...)`) as this is set up as a global constant.

**Comment by sferoze** [comment]
Its working now! Thanks so much for your patience in helping me get it work! The framework is a little like magic! I thought I would have to manually save everything into core data but it looks like it just works! This is such an awesome framework.

**Comment by martijnwalraven** [comment]
Thanks :) Hopefully I'll have more time to work on the documentation soon.

**Comment by katopz** [comment]
This look nice! any hit and run example or blog/doc for this yet?

**Comment by martijnwalraven** [comment]
Not yet, unfortunately. I plan on giving this project some more love after Xcode 7 comes out (hopefully around the same time I've finished some other work I'm doing). The idea is definitely to improve the documentation and also to write a blog post or tutorial.

**Comment by karlml** [comment]
Look forward to some more documentation :-)  While creating user and loving in is easy, I have spent hours trying to find out how to download a collection and how to get notified of updates.  I don't think i need the CoreData yet.

I added this in viewWillAppear:
        Meteor.addSubscriptionWithName("myCollectionName")


But then what?  When I add an entry to the collection, the is no console log.  How do I catch it?

**Comment by karlml** [comment]
For others that use swift, this is and example of how to subscribe to a collection and then print the entire collection:

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)

        Meteor.addSubscriptionWithName("myCollectionName", parameters: [3]) { (error) -> Void in
            if (error == nil) {
                print("success I have subscribet to myCollectionName")
            }else{
                print("error: \(error)")
            }
        }
        
        let myCollection = Meteor.database.collectionWithName("myCollectionName")
        let allMyCollection = myCollection.allDocuments
        print(allMyCollection)
    }
